File: linked-list/singly/singly.py
class SinglyLinkedList:
    def __init__(self) -> None:
        self.head = Node()
        self.tail = Node()
        self.head.next = self.tail
        
    def insert_new_head(self, node):
        tempNode = self.head
        self.head = node
        node.next = tempNode
        
    def insert_new_tail(self,node):
        self.tail.next = node
        self.tail = node
        
    def insert(self, prevNode, newNode):
        tempNode = prevNode.next
        prevNode.next = newNode
        newNode.next = tempNode
        
    def delete(self, node):
        if self.head == node:
            if node.next == self.tail: 
                print("Cannot delete Head or Tail without other nodes.")
                return
            self.head = node.next
            return
        
        prevNode = self.head
        while prevNode.next != node and prevNode.next is not None:
            prevNode = prevNode.next
        if prevNode.next == node:
            if node == self.tail:
                self.tail = prevNode
            prevNode.next = node.next
        else:
            print("Node not found")
        
    def printList(self):
        node = self.head
        while node:
            print(f"Node [{node.val}] -> ", end="")
            node = node.next
        print("None")
    
    def search(self, val):
        node = self.head
        while node.val != val and node.next is not None:
            node = node.next
        if node.val == val:
            return True
        return False
            
class Node:
    def __init__(self, val=0) -> None:
        self.val = val
        self.next = None
        
linkedList = SinglyLinkedList()
linkedList.delete(linkedList.head)
linkedList.insert(linkedList.head, Node(1))
linkedList.insert_new_head(Node(69))
linkedList.insert_new_tail(Node(20))
linkedList.printList()
linkedList.delete(linkedList.head.next.next)
linkedList.printList() # 69 -> 0 -> 0 -> None
linkedList.delete(linkedList.tail) 
linkedList.printList() # 69 -> 0 -> None
linkedList.delete(linkedList.head) # Error
linkedList.printList()  
linkedList.insert(linkedList.head, Node(100))
linkedList.delete(linkedList.head) 
linkedList.printList() # 100 -> 0 -> None
print(linkedList.search(100))

File: linked-list/singly/singly.ts
class SignlyLinkedList<T extends number> {
  head: LinkedListNode<T>;
  tail: LinkedListNode<T>;
  constructor() {
    this.head = new LinkedListNode<T>();
    this.tail = new LinkedListNode<T>();
    this.head.next = this.tail;
  }

  insertAtHead(node: LinkedListNode<T>) {
    const tempNode = this.head;
    this.head = node;
    node.next = tempNode;
  }

  insertNewTail(node: LinkedListNode<T>) {
    this.tail.next = node;
    this.tail = node;
  }

  insert(prevNode: LinkedListNode<T>, newNode: LinkedListNode<T>) {
    const tempNode = prevNode.next;
    prevNode.next = newNode;
    newNode.next = tempNode;
  }

  delete(node?: LinkedListNode<T>) {
    if (!node) {
      console.log("NaN, please pass in a Node.");
      return;
    }
    if (this.head == node) {
      if (node.next == this.tail) {
        console.log("Cannot delete Head or Tail without other nodes.");
        return;
      }
      this.head = <LinkedListNode<T>>node.next;
    }

    let prevNode = this.head;
    while (prevNode.next != node && prevNode.next) prevNode = prevNode.next;
    if (prevNode.next == node) {
      if (node == this.tail) {
        this.tail = prevNode;
      }
      prevNode.next = node.next;
    } else console.log("Node not found");
  }

  printList() {
    let node = this.head;
    let stringBuilder = "";
    while (node) {
      stringBuilder += `Node [${node.val}] -> `;
      node = <LinkedListNode<T>>node.next;
    }
    stringBuilder += "None";
    console.log(stringBuilder);
  }

  search(val: T) {
    let node = this.head;
    while (node.val !== val && node.next) node = node.next;
    node.val == val ? true : false;
  }
}

// class LinkedListNode<T extends number> {
//   next: LinkedListNode<T> | undefined = undefined;
//   constructor(public val?: T) {
//     this.val = val ? val : (0 as T);
//   }
// }

// const linkedList = new SignlyLinkedList<number>();
// linkedList.printList();
// linkedList.insert(linkedList.head, new LinkedListNode<number>(1));
// linkedList.insertAtHead(new LinkedListNode<number>(69));
// linkedList.printList(); // 69 -> 0 -> 1 -> 0
// linkedList.delete(linkedList.head.next?.next);
// linkedList.printList(); // 69 -> 0 -> 0 -> None
// linkedList.delete(linkedList.tail);
// linkedList.printList(); // 69 -> 0 -> None
// linkedList.delete(linkedList.head); // Error
// linkedList.printList();
// linkedList.insert(linkedList.head, new LinkedListNode<number>(100));
// linkedList.delete(linkedList.head);
// linkedList.printList(); // 100 -> 0 -> None
